pricegraph/src/orderbook/scalar.rs

```diff
@@ -169,3 +167,158 @@ fn assert_strictly_positive_and_finite(value: f64) -> f64 {
+/// chosen for exchange rate weights. Specifically, weights must be in a
+/// logarithmic scale so that adding them is equivalent to multiplying exchange
+/// rates. Since `log2` of the exchange rate is used, and the range of these
+/// exchange rates are:
+/// ```
+/// [MIN_AMOUNT / u128::MAX, u128::MAX / MIN_AMOUNT]
+/// ```
+///
+/// In the logarithmic scale, this range is:
+/// ```
+/// [-114.71, 114.71]
+/// ```
+///
+/// Furthermore, these weights can be added at most 2^16 times (this is the
+/// maximum number of tokens in the exchange), so the total range that must be
+/// representable by the magnitude bits is:
+/// ```
+/// [-7517784, 7517784]
+/// ```
+///
+/// This number fits in 23 bits. However, an additional bit is needed in order
+/// to have a special "infinite" value which is required by the Bellman-Ford
+/// implementation.
+///
+/// This leaves 104 fractional bits. Note that we want **as many fractional bits
+/// as possible** to keep as much precision as possible for values very close to
+/// `0.0`. With `104` fractional bits, we can represent without precision loss
+/// any `f64` that is greater than `1e-51` (since `-104 - 53 == 51` where `53 is
+/// the number of bits of precision in an `f64`.
+type Fixed24x104 = i128;
+
+/// The 24x104 fixed point number scaling factor.
+///
+/// Note that this value is a **power of 2** to make sure that multiplying with
+/// `f64`s does not cause precision issues.
+const FIXED_24X104_SCALING_FACTOR: f64 = (1u128 << 104) as _;
+
+/// An opaque weight for an exchange rate used by the pathfinding algorithm.
+///
+/// Internally, the weight is a represented as an fixed point number.
+#[derive(Clone, Copy, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Weight(Fixed24x104);
+
+impl Weight {
+    /// Creates a new graph weight from a floating point number.
+    pub fn new(value: f64) -> Self {
+        // TODO(nlordell): In the future, it might be nice to compute the `log2`
+        // already as a fixed point value in order to have more precision.
+        // Currently, since it is computed as a `f64` it is limited to 53 bits
+        // of precision instead of the full `104 + 7` that is possible given
+        // the range of the `log2` values and size of the fixed point number.
+
+        let weight = value.log2() * FIXED_24X104_SCALING_FACTOR;
+        debug_assert!(
+            -114.72 * FIXED_24X104_SCALING_FACTOR <= weight
+                && weight < 114.72 * FIXED_24X104_SCALING_FACTOR,
+        );
+
+        Weight(weight as _)
+    }
+}
+
+impl FloatMeasure for Weight {
+    fn zero() -> Self {
+        Weight(0)
+    }
+
+    fn infinite() -> Self {
+        // NOTE: Use a special marker value to represent +∞ which is needed by
+        // the `petgraph` Bellman-Ford implementation. `i128::MIN` is chosen so
+        // that the the range of non-infinite values are symmetric around `0`.
+        Weight(i128::MIN)
```

Comment:
Technically we can pick any value that has the left most bit set to 1, right? Just making sure I understand it correctly.

Reply:
No, leftmost bit is always 1 for negative numbers, this has to be a specific value. The reason I mentioned we "need 1 bit for ∞` is actually a little inaccurate and a bit of lie. We need it in order to have enough space for the two's complement of `7517784` while still reserving a special value for ∞.

Reply:
Updated the comment to make more sense.